### PR TITLE
fix: bind noexcept and ref-qualified methods from unregistered base classes

### PR DESCRIPTION
**Pull request by Skylion007**
## Description
Fix issue #2234 . Fix overloading to accept noexcept function types too and strip them. Mostly authored by ClaudeCode with a lot of human intervention.

fix: support noexcept and ref-qualified methods from unregistered base classes (issue #2234)
## Problem

In C++17, `noexcept` became part of the function type (P0012R1). This means
`int (Base::*)() const noexcept` and `int (Base::*)() const` are **distinct
types**. When a derived class inherits methods from an unregistered base,
`method_adaptor` must recast the member function pointer from `Base::*` to
`Derived::*`. Without explicit template specializations for noexcept (and
ref-qualified) variants, the generic `F&&` fallback was selected, leaving the
pointer typed as `Base::*` — causing pybind11 to look up `Base` as `self`,
which fails at runtime for unregistered base classes.

Similarly, `cpp_function` lacked constructors for `noexcept`, `& noexcept`,
`const noexcept`, `const & noexcept`, `&&`, `const &&`, `&& noexcept`, and
`const && noexcept` member function pointer types, so those overloads silently
fell through to the generic lambda constructor.

### Solution

**`include/pybind11/pybind11.h`**

- Add `cpp_function` constructors for all missing qualifier combinations:
  - `&&`, `const &&` (unconditional): use `std::move(*c).*f` in the lambda
  - `noexcept`, `const noexcept`, `& noexcept`, `const & noexcept`,
    `&& noexcept`, `const && noexcept` (under `#ifdef __cpp_noexcept_function_type`)
- Add `detail::rebind_member_ptr<Derived, T>` type trait mapping
  `Return (Base::*)(Args...) qualifiers` to `Return (Derived::*)(Args...) qualifiers`
  across all 12 qualifier combinations, generated via a local
  `PYBIND11_REBIND_MEMBER_PTR` macro.
- Add `detail::adapt_member_ptr<Derived>(pmf)` — shared `constexpr` helper
  called by all `method_adaptor` overloads: runs the `is_accessible_base_of`
  static_assert then returns the implicitly-cast pointer.
- Replace the ad-hoc `method_adaptor` overloads with a `PYBIND11_METHOD_ADAPTOR`
  macro covering all 11 qualified variants (plus the no-qualifier overload
  written explicitly to avoid MSVC C4003). All overloads are `constexpr`.
- Add unconditional `&` / `const &` `method_adaptor` overloads (no `#ifdef`
  guard) to fix C++14 builds where `noexcept` is stripped from the type but
  the `&` qualifier is not.

**`include/pybind11/numpy.h`**

- Add `vectorize` overloads for `noexcept` free function pointers.

**Tests**

- `test_methods_and_attributes`: `NoexceptDerived` / `NoexceptUnregisteredBase`
  (C++17 noexcept from unregistered base), `RValueRefDerived` /
  `RValueRefUnregisteredBase` (`&&`/`const &&` methods that move-out state to
  confirm `std::move(*c).*f` semantics), `NoexceptOverloaded` (`overload_cast`
  with noexcept), static_asserts verifying `method_adaptor` preserves all
  qualifier combinations.
- `test_buffers`: `def_buffer` with `noexcept` member function pointers.
- `test_numpy_vectorize`: `vectorize` with `noexcept` free function pointers.

### Compatibility

- No behaviour change for existing code — all new `cpp_function` constructors
  and `method_adaptor` overloads are additive and selected only for the new
  type signatures.
- C++14 compatible (noexcept constructors/overloads gated on
  `__cpp_noexcept_function_type`).
- MSVC C4003 avoided by writing the no-qualifier specializations explicitly
  rather than invoking macros with an empty argument.
- clang-tidy `bugprone-macro-parentheses` suppressed with `NOLINTBEGIN` /
  `NOLINTEND` around the macro definitions (the `qualifiers` argument appears
  in type position where parenthesizing it would be invalid C++).

## Suggested changelog entry:

* Fix binding of ``noexcept`` and ref-qualified (``&``, ``&&``) methods
  inherited from unregistered base classes in C++17, where ``noexcept`` is
  part of the function type. ``method_adaptor`` now has specializations for
  all qualifier combinations (``const``, ``&``, ``const &``, ``&&``,
  ``const &&``, and their ``noexcept`` variants), and ``cpp_function`` gains
  matching constructors. ``&&``-qualified methods correctly use
  ``std::move(*self).*f`` in the generated lambda.

### CODE REVIEW

File: include/pybind11/pybind11.h [hunk 1]
```diff
@@ -1905,6 +1947,61 @@ auto method_adaptor(Return (Class::*pmf)(Args...) const) -> Return (Derived::*)(
...
+    -> Return (Derived::*)(Args...) & noexcept {
+    static_assert(
+        detail::is_accessible_base_of<Class, Derived>::value,
+        "Cannot bind an inaccessible base class method; use a lambda definition instead");
+    return pmf;
+}
+
+template <typename Derived, typename Return, typename Class, typename... Args>
+auto method_adaptor(Return (Class::*pmf)(Args...) const & noexcept)
+    -> Return (Derived::*)(Args...) const & noexcept {
```
Comment (Skylion007) [inline include/pybind11/pybind11.h]: I think I need to add tests for '&&' here i guess. Otherwise the fall through the perfect forwarding template... Sigh... I'll try construct some tests later I guess.

File: include/pybind11/pybind11.h [hunk 2]
```diff
@@ -1905,6 +1947,61 @@ auto method_adaptor(Return (Class::*pmf)(Args...) const) -> Return (Derived::*)(
     return pmf;
 }
 
+template <typename Derived, typename Return, typename Class, typename... Args>
```
Comment (Skylion007) [inline include/pybind11/pybind11.h:2000]: Maybe able to massively simplify this with a macro...

File: include/pybind11/detail/common.h [hunk 3]
```diff
@@ -1056,14 +1056,30 @@ struct strip_function_object {
...
+#endif
+template <typename T>
+using remove_noexcept_t = typename remove_noexcept<T>::type;
+
 // Extracts the function signature from a function, function pointer or lambda.
+// Strips noexcept from the result so that factory/pickle_factory partial specializations
+// (which match plain Return(Args...)) work correctly with noexcept callables (issue #2234).
 template <typename Function, typename F = remove_reference_t<Function>>
-using function_signature_t = conditional_t<
+using function_signature_t = remove_noexcept_t<conditional_t<
```
Comment (Skylion007) [inline include/pybind11/detail/common.h:1086]: Should I normalize it here for all pybind11? Or keep it as utility that and create a new alias that always strips the noexcept

Comment (Skylion007) [reply include/pybind11/detail/common.h:1086]: on the flip side we could use the noexcept tag handling to optimize away exception handling for certain edge cases like def_buffer maybe?

File: tests/test_methods_and_attributes.cpp [hunk 4]
```diff
@@ -161,6 +161,68 @@ class RegisteredDerived : public UnregisteredBase {
...
+    int m_value = 99;
+};
+class NoexceptDerived : public NoexceptUnregisteredBase {
+public:
+    using NoexceptUnregisteredBase::NoexceptUnregisteredBase;
+};
+
+// Exercises cpp_function(Return (Class::*)(Args...) &&, ...) and
+// cpp_function(Return (Class::*)(Args...) const &&, ...) via an unregistered base.
+class RValueRefUnregisteredBase {
```
Comment (Skylion007) [inline tests/test_methods_and_attributes.cpp:194]: Also unsure if should encourage rvalue overloads, but they were always supported by manually calling them via lambdas sooo

**Comment by rwgk** [comment]
Hi @Skylion007 could you please validate my high-level understanding of this PR?

This is the result of me chatting with Cursor (GPT-5.3 Codex High) for a bit:

___

- **Status quo on current master is inconsistent/partial.**
  - Some `noexcept` callables bind today (often simpler signatures).
  - Others fail to bind cleanly, or fall through less-specific template paths.
  - In some cases (notably around unregistered-base/member-pointer adaptation), binding may compile but not behave correctly at runtime.

- **Throwing through bound `noexcept` functions remains process-terminating by C++ semantics.**
  - If a C++ exception escapes a `noexcept` function, `std::terminate` is triggered.
  - This preempts pybind11 exception translation (`error_already_set` -> Python exception).
  - Therefore, Python-visible calls may abort the process if a bound `noexcept` function throws.

- **What PR 5992 improves.**
  - It broadens/normalizes qualifier handling (`const`, `&`, `&&`, and `noexcept` combinations), making signature matching more uniform and predictable.
  - It reduces accidental fallback paths and fixes specific C++17 `noexcept`-typed member-pointer adaptation issues (issue #2234 context).
  - It does not change the underlying C++ runtime rule for exceptions escaping `noexcept`.

**Comment by rwgk** [comment]
Hi @Skylion007 I added five Cursor-generated commits (all using GPT-5.3 Codex High). I'll do a manual review here after I see that the CI is still passing.

**Comment by Skylion007** [comment]
So I'm new sure if all those Codex commits were necessary due to the strip_except function_signature utility I added? Can you verify those tests fail on the previous commit? And you don't always need specific lvalue and rvalue overload handling.

File: include/pybind11/pybind11.h [hunk 5]
```diff
@@ -2361,6 +2508,40 @@ class class_ : public detail::generic_type {
         return def_buffer([func](const type &obj) { return (obj.*func)(); });
     }
 
+    // Intentionally no &&/const&& overloads: buffer protocol callbacks are invoked on an
```
Comment (Skylion007) [inline include/pybind11/pybind11.h:2529]: Valid

**Comment by rwgk** [comment]
> So I'm new sure if all those Codex commits were necessary due to the strip_except function_signature utility I added? Can you verify those tests fail on the previous commit? And you don't always need specific lvalue and rvalue overload handling.

Good idea. Will do (probably this weekend).

File: include/pybind11/pybind11.h [hunk 6]
```diff
@@ -1880,29 +1970,86 @@ inline void add_class_method(object &cls, const char *name_, const cpp_function
...
+        using type = Return (Derived::*)(Args...) qualifiers;                                     \
+        using source_class = Class;                                                               \
+    }
+PYBIND11_REBIND_MEMBER_PTR(const);
+PYBIND11_REBIND_MEMBER_PTR(&);
+PYBIND11_REBIND_MEMBER_PTR(const &);
+PYBIND11_REBIND_MEMBER_PTR(&&);
+PYBIND11_REBIND_MEMBER_PTR(const &&);
+#ifdef __cpp_noexcept_function_type
+PYBIND11_REBIND_MEMBER_PTR(noexcept);
```
Comment (Skylion007) [inline include/pybind11/pybind11.h:2018]: I didn't do the overloads for volatile because I wasn't sure if we would should even support them here @rwgk

File: include/pybind11/pybind11.h [hunk 7]
```diff
@@ -1880,29 +1970,86 @@ inline void add_class_method(object &cls, const char *name_, const cpp_function
...
+                        int>
+    = 0>
+constexpr auto method_adaptor(F &&f) -> decltype(std::forward<F>(f)) {
+    return std::forward<F>(f);
+}
+
+template <typename Derived,
+          typename T,
+          typename Adapted = typename detail::rebind_member_ptr<Derived, T>::type>
+constexpr Adapted method_adaptor(T pmf) {
```
Comment (Skylion007) [inline include/pybind11/pybind11.h:2065]: Probably should add a static assert here that T is_member_ptr for sanity.

**Comment by rwgk** [comment]
FYI — I'm finally getting back here. Trying to catch up now, and looking into "So I'm new sure if all those Codex commits were necessary"

**Comment by rwgk** [comment]
> So I'm new sure if all those Codex commits were necessary due to the strip_except function_signature utility I added? Can you verify those tests fail on the previous commit? And you don't always need specific lvalue and rvalue overload handling.

I don't have the free bandwidth to mentally sort through the jungle of C++ rules myself, but I used

Cursor Opus 4.6 1M Thinking

(different from what I used before) to go through a couple experiments. The Cursor-generated report is below. My final prompt was:

>   Could you please confirm that my understanding is correct: Under the constraint that we don't want to remove any tests, new or pre-existing, all added overloads are strictly needed?

Response:

Yes, that's correct. Experiment 1 already demonstrated this directly: applying all the Codex test additions to Skylion007's code (without any library changes) fails to compile all three test files. Every library overload added by the Codex commits is exercised by at least one test, and removing any overload while keeping its test causes a compile error.

___

## Experiment 1: Apply only Codex test additions to Skylion007's code

Checked out Skylion007's last commit (`e4ece99`), applied *only* the test additions from the 5 rwgk/Codex commits (`798e516..cd6ef9d`), and attempted to build (GCC, C++20, scons).

All three test files failed to compile:

### 1. `test_buffers.cpp` — 4 compile errors

`def_buffer` with `&`, `const &`, `& noexcept`, and `const & noexcept` member function pointers all fail. The existing `def_buffer` template only had overloads for unqualified and `const` (plus `noexcept` and `const noexcept`). The `&`-qualified pointers match the generic template but the generated lambda tries to call `func(...)` instead of `(obj.*func)()`, causing "must use '.\*' or '->\*'" errors.

### 2. `test_methods_and_attributes.cpp` — ~24 compile errors

`overload_cast_impl` had no `operator()` overloads for `&`, `const &`, `&&`, `const &&` (or their noexcept combinations) member function pointers. The `static_assert`s and `.def()` calls using `overload_cast` with ref-qualified methods all fail with "no match for call to overload_cast_impl".

### 3. `test_numpy_vectorize.cpp` — 4 compile errors

`vectorize()` had no overloads for `&`, `const &`, `& noexcept`, or `const & noexcept` member function pointers. "no matching function for call to 'vectorize'" for all four qualifier combinations.

## Why `remove_noexcept_t` / `function_signature_t` doesn't cover these cases

Skylion007's `remove_noexcept_t` in `function_signature_t` strips `noexcept` from the function *signature type* used for `cpp_function` constructor matching. That works well for the `cpp_function` path. But it doesn't help with:

- **`def_buffer`** — uses explicit member-function-pointer template overloads (not `function_signature_t`)
- **`vectorize`** — same, uses explicit overloads
- **`overload_cast_impl`** — same, uses explicit `operator()` overloads

All three of these APIs needed new overloads for the **lvalue-ref (`&`, `const &`)** qualifier combinations (and their `noexcept` variants). These are ref-qualifier gaps, not noexcept issues.

## Experiment 2: Fine-grained analysis of `overload_cast_impl` overloads

The Codex commits added four pairs of `overload_cast_impl` overloads in `common.h`:
`&`, `&&`, `& noexcept`, `&& noexcept` (each as a non-const/const pair).

To test whether any of these are unnecessary, I removed all four pairs from `overload_cast_impl` in `common.h` while keeping all Skylion007 code *and* the rwgk `def_buffer`/`vectorize` library changes. I reverted all test files to Skylion007's versions (removing the `RefQualifiedOverloaded` tests added by the Codex commits).

**Result: builds and passes all 1300 Skylion007 tests cleanly** (the 7 missing vs HEAD's 1307 are exactly the Codex-added `RefQualifiedOverloaded`/ref-qualified buffer/vectorize tests).

This means the `overload_cast_impl` `&`/`&&` overloads are **self-contained additions** — they are only exercised by tests that were themselves added in the same Codex commits. No pre-existing code depends on them.

I also verified the individual pairs:
- Removing only `&&`/`const &&` (keeping `&`/`const &`): only the new `RefQualifiedOverloaded` `&&` tests break.
- Removing only `&`/`const &` (keeping `&&`/`const &&`): only the new `RefQualifiedOverloaded` `&` tests break.

## Summary: what's necessary vs what's consistency-only

| Change | Necessary for new capability? | Notes |
|--------|-------------------------------|-------|
| `def_buffer` `&`/`const &` overloads (`pybind11.h`) | **Yes** | Without them, `&`-qualified member function pointers fail to compile in `def_buffer` |
| `def_buffer` `& noexcept`/`const & noexcept` overloads | **Yes** | Same, for `& noexcept` combinations |
| `vectorize` `&`/`const &` overloads (`numpy.h`) | **Yes** | Same pattern as `def_buffer` |
| `vectorize` `& noexcept`/`const & noexcept` overloads | **Yes** | Same |
| `overload_cast_impl` `&`/`const &` pair (`common.h`) | Consistency | Makes `overload_cast` work with `&`-qualified methods (which `def_buffer`/`vectorize` now support). Only tested by Codex-added tests. |
| `overload_cast_impl` `&&`/`const &&` pair (`common.h`) | Consistency | `def_buffer`/`vectorize` intentionally omit `&&`, so this only matters for `.def()` + `overload_cast`. Only tested by Codex-added tests. |
| Noexcept variants of the above `overload_cast_impl` pairs | Same as their non-noexcept counterparts | |

The `def_buffer` and `vectorize` overloads are genuinely necessary — they enable new capability for ref-qualified member function pointers that couldn't work before.

The `overload_cast_impl` overloads are forward-looking consistency additions. They don't fix any existing bug, but they prevent `overload_cast` from being a surprising gap when users try to disambiguate ref-qualified methods. The `&&`/`const &&` pair is the most "optional" of these, since `def_buffer`/`vectorize` intentionally don't support `&&`. However, Skylion007's own commits added `cpp_function` constructors for `&&`-qualified methods, so `overload_cast` support for those is a reasonable completeness addition for use with `.def()`.

File: include/pybind11/pybind11.h (hunk 6 above)
Comment (rwgk) [reply include/pybind11/pybind11.h:2018]: That sounds good to me. Actual use cases are probably exceedingly rare.

File: include/pybind11/pybind11.h (hunk 7 above)
Comment (rwgk) [reply include/pybind11/pybind11.h:2065]: Done: commit 1d83b6f41b9c8530d09b53d086b0c293141ddae3
